aspnetcore/web-api/error-handling.md

```diff
@@ -0,0 +1,230 @@
+    
+1. Apply attribute routing to controller actions:
+
+    ```csharp
+    [ApiController]
+    public class ErrorController : ControllerBase
+    {
+        [Route("/error-local-development")]
+        public IActionResult ErrorLocalDevelopment(
+            [FromServices] IWebHostEnvironment webHostEnvironment)
+        {
+            if (!webHostEnvironment.IsDevelopment())
+            {
+                throw new InvalidOperationException(
+                    "This shouldn't be invoked in non-development environments.");
+            }
+    
+            var context = HttpContext.Features.Get<IExceptionHandlerFeature>();
+            return Problem(
+                detail: context.Error.StackTrace,
+                title: context.Error.Message);
+        }
+    
+        [Route("/error")]
+        public IActionResult Error() => Problem();
+    }
+    ```
+
+## Use exceptions to modify the response
+
+You may want to modify the contents of the response from outside of the controller. In ASP.NET 4.x Web API, one way to do this was using the `HttpResponseException` type. ASP.NET Core doesn't include an equivalent type. Support for `HttpResponseException` can be added with the following steps:
+
+1. Create a well-known exception type named `HttpResponseException`:
+
+    ```csharp
+    public class HttpResponseException : Exception
+    {
+        public int Status { get; set; } = 500;
+    
+        public object Value { get; set; }
+    }
+    ```
+
+1. Create an action filter named `HttpResponseExceptionFilter`:
+
+    ```csharp
+    public class HttpResponseExceptionFilter : IActionFilter, IOrderedFilter
+    {
+        public int Order { get; set; } = int.MaxValue - 10;
+    
+        public void OnActionExecuting(ActionExecutingContext context) {}
+    
+        public void OnActionExecuted(ActionExecutedContext context)
+        {
+            if (context.Exception is HttpResponseException exception)
+            {
+                context.Result = new ObjectResult(exception.Value)
+                {
+                    Status = exception.Status,
+                };
+                context.ExceptionHandled = true;
+            }
+        }
+    }
+    ```
+
+1. In `Startup.ConfigureServices`, add the action filter to the filters collection:
+
+    ```csharp
+    public void ConfigureServices(IServiceCollection services)
+    {
+        services.AddControllers(options => 
+            options.Filters.Add(new HttpResponseExceptionFilter()));
+    }
+    ```
+
+## Validation failure error response
+
+For web API controllers, MVC responds with a `ValidationProblemDetails` response type when model validation fails. MVC uses the results of <xref:Microsoft.AspNetCore.Mvc.ApiBehaviorOptions.InvalidModelStateResponseFactory> to construct
+the error response for a validation failure. The following example uses the factory to change the default response type to `SerializableError`:
```

Comment:
Remove this line ... there's a return stuck on the end of Line 163. I've already compensated for the text on Line 164 by adding it to my suggestion for Line 163.

```suggestion
```